### PR TITLE
feat: allow configuring timeout and retries for upstream with ingress

### LINKED ISSUE #1783: request help: native Kubernetes ingress  yaml  add  timeout config
### Issue description

image

how to change default  timeout and  retry config in native Kubernetes ingress  yaml 

### Environment

- your apisix-ingress-controller version (output of apisix-ingress-controller version --long):
- your Kubernetes cluster version (output of kubectl version):
- if you run apisix-ingress-controller in Bare-metal environment, also show your OS version (uname -a):

### PR DESCRIPTION
**Pull request by Revolyssup**
### Type of change:
Fixes #1783 

Documentation to be updated.

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches
- [ ] Documentation
- [ ] Refactor
- [ ] Chore
- [ ] CI/CD or Tests

### Pre-submission checklist:

- [ ] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

**Comment by lingsamuel** [comment]
`make update-gofmt` (or `make update-all`) is required

**Comment by tao12345666333** [comment]
yes, we need an e2e test case to cover this one.

**Comment by Revolyssup** [comment]
@tao12345666333 How do you  suggest I simulate backend for testing retries and timeouts? For retries, I configured an http server on port 8080 that responds with 200 on 3rd attempt and the I set the retry to 3, but tests seem to fail. Do you have a better approach?

**Comment by tao12345666333** [comment]
Using the serverless plugin of APISIX, it will be easier to simulate with Lua.

**Review by Sn0rt — APPROVED** [review]

**Comment by AlinsRan** [comment]
Please provide documentation for this feature.
https://apisix.apache.org/docs/ingress-controller/concepts/annotations

File: pkg/providers/ingress/translation/annotations/types.go [hunk 1]
```diff
@@ -27,6 +27,12 @@ const (
 	AnnotationsEnableWebSocket  = AnnotationsPrefix + "enable-websocket"
 	AnnotationsPluginConfigName = AnnotationsPrefix + "plugin-config-name"
 	AnnotationsUpstreamScheme   = AnnotationsPrefix + "upstream-scheme"
+
+	//support retries and timeouts on upstream
+	AnnotationsUpstreamRetry          = AnnotationsPrefix + "retry"
```
Comment (AlinsRan) [inline pkg/providers/ingress/translation/annotations/types.go:32]: I suggest using  `retries`

File: test/e2e/suite-annotations/upstreamretry.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+var _ = ginkgo.Describe("suite-annotations: annotations.networking/v1 upstream", func() {
+	s := scaffold.NewDefaultScaffold()
+	ginkgo.It("Test timeout: 1 retry and long timeout", func() {
+		ing := `
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  annotations:
+    k8s.apisix.apache.org/retry: "1"
+    k8s.apisix.apache.org/timeout.read: "20s"
```
[the remark this answers is not in the document]
Comment (Revolyssup) [reply test/e2e/suite-annotations/upstreamretry.go:37]: How about, `upstream.retries` and `upstream.timeout.read`?

File: docs/en/latest/concepts/annotations.md [hunk 3]
```diff
@@ -413,3 +413,63 @@ spec:
                 port:
                   number: 80
 ```
+
+## Upstream retries
+
+This annotation can be used to configure retries among multiple nodes in an upstream. You may want the proxy to retry when requests occur faults like transient network errors or service unavailable, by default the retry count is 1. You can change it by specifying the retries field.
```
Comment (kayx23) [inline docs/en/latest/concepts/annotations.md:419]:
```suggestion
This annotation can be used to configure retries among multiple nodes in an upstream. You may want the proxy to retry when requests occur faults like transient network errors or service unavailable. By default the retry count is 1. You can change it by specifying the retries field.
```

File: docs/en/latest/concepts/annotations.md [hunk 4]
```diff
@@ -413,3 +413,63 @@ spec:
...
+                name: httpbin
+                port:
+                  number: 80
+```
+
+## Upstream timeout
+
+This annotation can be used to configure different types of timeout on an upstream. The default connect, read and send timeout are 60s, which might not be proper for some applications, just change them in the timeout field.
+
+The below example sets the connect, read and timeout to 5s, 10s, 10s respectively.
```
Comment (kayx23) [inline docs/en/latest/concepts/annotations.md:451]: `connect, read and timeout` : is this not `retries, read, and connect timeout`?

File: docs/en/latest/concepts/annotations.md [hunk 5]
```diff
@@ -413,3 +413,63 @@ spec:
...
+            backend:
+              service:
+                name: httpbin
+                port:
+                  number: 80
+```
+
+## Upstream timeout
+
+This annotation can be used to configure different types of timeout on an upstream. The default connect, read and send timeout are 60s, which might not be proper for some applications, just change them in the timeout field.
```
Comment (kayx23) [inline docs/en/latest/concepts/annotations.md:449]:
```suggestion
This annotation can be used to configure different types of timeout on an upstream. The default connect, read and send timeout are 60s, which might not be proper for some applications.
```

**Review by kayx23 — COMMENTED** [review]
some suggestions for docs

File: docs/en/latest/concepts/annotations.md (hunk 5 above)
Comment (Revolyssup) [reply docs/en/latest/concepts/annotations.md]: done

File: docs/en/latest/concepts/annotations.md [hunk 6]
```diff
@@ -413,3 +413,63 @@ spec:
                 port:
                   number: 80
 ```
+
+## Upstream retries
+
+This annotation can be used to configure retries among multiple nodes in an upstream. You may want the proxy to retry when requests occur faults like transient network errors or service unavailable, by default the retry count is 1. You can change it by specifying the retries field.
+
+The following configuration configures the retries to 3, which indicates there'll be at most 3 requests sent to Kubernetes service httpbin's endpoints.
```
[the remark this answers is not in the document]
Comment (Revolyssup) [reply docs/en/latest/concepts/annotations.md:421]: fixed

File: docs/en/latest/concepts/annotations.md (hunk 4 above)
Comment (Revolyssup) [reply docs/en/latest/concepts/annotations.md]: fixed

File: test/e2e/suite-annotations/upstreamretry.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlinsRan) [reply test/e2e/suite-annotations/upstreamretry.go]: I suggest using - as the separator for annotations name, which is the format we are currently using. Please maintain consistency.

`upstream-retries` 
`upstream-read-timeout`.

**Review by AlinsRan — APPROVED** [review]

**Review by Sn0rt — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]

**Pull request merged**